### PR TITLE
journal: modernize journal ratelimiting

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/journal/journald-rate-limit.c [hunk 1]
```diff
@@ -23,137 +18,128 @@ static const int priority_map[] = {
...
+                return -ENOMEM;
 
         g->id = strdup(id);
         if (!g->id)
-                goto fail;
-
-        g->hash = siphash24_string(g->id, r->hash_key);
+                return -ENOMEM;
 
         g->interval = interval;
```
Comment (YHNdnzj) [inline src/journal/journald-rate-limit.c]: Use designated initializer for this?

File: src/journal/journald-rate-limit.c [hunk 2]
```diff
@@ -26,82 +21,43 @@ static const int priority_map[] = {
...
-                if (g->parent->lru_tail == g)
-                        g->parent->lru_tail = g->lru_prev;
-
-                LIST_REMOVE(lru, g->parent->lru, g);
-                LIST_REMOVE(bucket, g->parent->buckets[g->hash % BUCKETS_MAX], g);
-
-                g->parent->n_groups--;
-        }
+        if (g->groups_by_id && g->id)
+                ordered_hashmap_remove(g->groups_by_id, g->id);
```
Comment (YHNdnzj) [inline src/journal/journald-rate-limit.c]: Maybe, `assert_se(ordered_hashmap_remove_value(...))`?

File: src/journal/journald-rate-limit.c [hunk 3]
```diff
@@ -23,137 +18,128 @@ static const int priority_map[] = {
...
-void journal_ratelimit_free(JournalRateLimit *r) {
-        assert(r);
+DEFINE_TRIVIAL_CLEANUP_FUNC(JournalRateLimitGroup*, journal_ratelimit_group_free);
 
-        while (r->lru)
-                journal_ratelimit_group_free(r->lru);
-
-        free(r);
-}
+DEFINE_HASH_OPS_WITH_VALUE_DESTRUCTOR(
```
Comment (YHNdnzj) [inline src/journal/journald-rate-limit.c]:
```suggestion
DEFINE_PRIVATE_HASH_OPS_WITH_VALUE_DESTRUCTOR(
```

**Review by YHNdnzj — APPROVED** [review]

**Comment by yuwata** [comment]
@YHNdnzj Thank you for the review. Applied the above suggestions. Upgrading the green label.

File: src/journal/journald-rate-limit.c (hunk 2 above)
Comment (yuwata) [reply src/journal/journald-rate-limit.c]: `assert_se()` does not work here, as the entry is already removed if this is called from the destructor of the hashmap.

**Pull request merged**